File: s3RubySDKExample/sample/acl_serial_testing.rb
def acl_serial_testing(s3, buckets, creds)
  # create a bucket
  resp = s3.create_bucket({
    #  acl: "public-read", # accepts private, public-read, public-read-write, authenticated-read
    bucket: buckets[0], # required
  })

  # get bucket acl
  resp = s3.get_bucket_acl({
    bucket: buckets[0], # required
  })

  # list bucket acl information
  puts "receive from get bucket acl"
  puts "owner name " + resp.owner.display_name #=> String
  puts "owner id " + resp.owner.id #=> String
  #  puts resp.grants #=> Array
  puts "grantee name " + resp.grants[0].grantee.display_name #=> String
  #  puts "grantee email "+ resp.grants[0].grantee.email_address #=> String
  puts "grantee id " + resp.grants[0].grantee.id #=> String
  puts "grantee type " + resp.grants[0].grantee.type #=> String, one of "CanonicalUser", "AmazonCustomerByEmail", "Group"
  #  puts "grantee uri "+resp.grants[0].grantee.uri #=> String
  puts "grantee permission " + resp.grants[0].permission #=> String, one of "FULL_CONTROL", "WRITE", "WRITE_ACP", "READ", "READ_ACP"

  #Specify a canned ACL, or
  #Specify the permission for each grantee explicitly
  #Specify a canned ACL
  s3.put_bucket_acl({
    acl: "public-read", # accepts private, public-read, public-read-write, authenticated-read
    bucket: buckets[0], # required
  })

  s3.put_bucket_acl({
    bucket: buckets[0],
    access_control_policy: {
      grants: [
        {
          grantee: {
            type: "Group", # required, accepts CanonicalUser, AmazonCustomerByEmail, Group
            uri: "http://acs.amazonaws.com/groups/s3/LogDelivery",
          },
          permission: "WRITE_ACP", # accepts FULL_CONTROL, WRITE, WRITE_ACP, READ, READ_ACP
        },
        {
          grantee: {
            type: "CanonicalUser", # required, accepts CanonicalUser, AmazonCustomerByEmail, Group
            id: creds["Owner ID"],
          },
          permission: "FULL_CONTROL", # accepts FULL_CONTROL, WRITE, WRITE_ACP, READ, READ_ACP
        },
      ],
      owner: {
        display_name: creds["Owner Name"],
        id: creds["Owner ID"],
      },
    },
  })

  # get bucket acl
  resp = s3.get_bucket_acl({
    bucket: buckets[0], # required
  })

  # list bucket acl information
  puts "receive from get bucket acl"
  puts "owner name " + resp.owner.display_name #=> String
  puts "owner id " + resp.owner.id #=> String
  #  puts resp.grants #=> Array
  #  puts "grantee name " + resp.grants[0].grantee.display_name #=> String
  #  puts "grantee email "+ resp.grants[0].grantee.email_address #=> String
  #  puts "grantee id "+ resp.grants[0].grantee.id #=> String
  puts "grantee type " + resp.grants[0].grantee.type #=> String, one of "CanonicalUser", "AmazonCustomerByEmail", "Group"
  #  puts "grantee uri "+resp.grants[0].grantee.uri #=> String
  puts "grantee permission " + resp.grants[0].permission #=> String, one of "FULL_CONTROL", "WRITE", "WRITE_ACP", "READ", "READ_ACP"

  resp = s3.put_object({
    body: "Hello World!",
    bucket: buckets[0],
    key: "hello",
  })

  s3.put_object_acl({
    bucket: buckets[0],
    key: "hello", # required
    access_control_policy: {
      grants: [
        {
          grantee: {
            id: creds["Grantee Canonical ID"],
            type: "CanonicalUser", # required, accepts CanonicalUser, AmazonCustomerByEmail, Group
          },
          permission: "FULL_CONTROL", # accepts FULL_CONTROL, WRITE, WRITE_ACP, READ, READ_ACP
        },
        {
          grantee: {
            type: "CanonicalUser", # required, accepts CanonicalUser, AmazonCustomerByEmail, Group
            id: creds["Owner ID"],
          },
          permission: "FULL_CONTROL", # accepts FULL_CONTROL, WRITE, WRITE_ACP, READ, READ_ACP
        },
      ],
      owner: {
        display_name: creds["Owner Name"],
        id: creds["Owner ID"],
      },
    },
  })

  resp = s3.get_object_acl({
    bucket: buckets[0], # required
    key: "hello", # required
  })
  puts "----------------------------------------------------"
  puts "receive from get object acl"
  puts resp.owner.display_name #=> String
  puts resp.owner.id #=> String
  puts resp.grants[0].grantee.display_name #=> String
  puts resp.grants[0].grantee.id #=> String

  cleanup(s3, buckets)
end

File: s3JavaScriptSDK/sample/ACLSerialTesting.js
//1. put bucket ACL for canned public-read
//2. get bucket ACL

function aclSerialTest (){

	return Q.fcall(function(){
	var deferred = Q.defer();
	console.log('-----------------------------------------');
	console.log('testing acl begin!');
	deferred.resolve();
	return deferred.promise;
	})
	.then(function (){
		var params = {
				  Bucket: testBucketName, /* required */
				  ACL: 'public-read', //private | public-read | public-read-write | authenticated-read',
				};
				return putBucketAcl(params, s3);
	})
	.then(function (){
		return getBucketAcl (testBucketName,s3);
	})
	.then(function(){
		var params = {
				  Bucket: testBucketName, /* required */
				  AccessControlPolicy: {
				    Grants: [
				      {
				        Grantee: {
				          Type: 'Group', /* required */
				          URI: 'http://acs.amazonaws.com/groups/s3/LogDelivery'
				        },
				        Permission: 'WRITE_ACP',//'FULL_CONTROL | WRITE | WRITE_ACP | READ | READ_ACP'
				      },
				      {
				        Grantee: {
				          Type: 'Group', /* required */
				          URI: 'http://acs.amazonaws.com/groups/s3/LogDelivery'
				        },
				        Permission: 'WRITE'//'FULL_CONTROL | WRITE | WRITE_ACP | READ | READ_ACP'
				      },
				      {
				        Grantee: {
				          Type: 'Group', /* required */
				          URI: 'http://acs.amazonaws.com/groups/s3/LogDelivery'
				        },
				        Permission: 'READ_ACP'//'FULL_CONTROL | WRITE | WRITE_ACP | READ | READ_ACP'
				      },
					  {
						  Grantee: {
							  Type: 'CanonicalUser', /* required */
							  ID: ownerID
						  },
						  Permission: 'FULL_CONTROL',//'FULL_CONTROL | WRITE | WRITE_ACP | READ | READ_ACP'
					  },
				    ],
				    Owner: {
				      DisplayName: ownerName,
				      ID: ownerID
				    }
				  },
				};
		return putBucketAcl(params, s3);
	})
	.then(function (){
		return getBucketAcl (testBucketName,s3);
	})
	.then(function (){
	var params = {
			Bucket: testBucketName,
			Key: 'objectkey1',
			ACL: 'public-read',
			Body: 'object_data_1',
			ContentType :'text',
			}
	return s3Perform.putObject(params, s3);
	})
	.then(function (){
		var params = {
		  Bucket: testBucketName, /* required */
		  Key: 'objectkey1', /* required */
		  AccessControlPolicy: {
		    Grants: [
		      {
		        Grantee: {
		          Type: 'CanonicalUser',
		          ID: granteeId,
		        },
		        Permission: 'FULL_CONTROL'
		      },
				{
					Grantee: {
						Type: 'CanonicalUser', /* required */
						ID: ownerID
					},
					Permission: 'FULL_CONTROL',//'FULL_CONTROL | WRITE | WRITE_ACP | READ | READ_ACP'
				}
		    ],
		    Owner: {
		      DisplayName: ownerName,
		      ID: ownerID
		    }
		  },
		};
		return putObjectAcl(params, s3);
	})
	.then(function (){
		var params = {
				  Bucket: testBucketName, /* required */
				  Key: 'objectkey1',
		};
		return getObjectAcl(params, s3);
	})
	.then(function(){
		return deleteAllObjects(s3, testBucketName);
	});
}


function putBucketAcl(params, s3)
{
	var deferred = Q.defer();
	s3.putBucketAcl(params, function(err, data) {
		  if (err) console.log(err, err.stack); // an error occurred
		  else {
		  	console.log('put bucket acl succeed');
		  	deferred.resolve(data);
		  }           // successful response
	});
	return deferred.promise;
}

function getBucketAcl (BucketName,s3){
	var deferred = Q.defer();
	var params = {
	  Bucket: BucketName /* required */
	};
	s3.getBucketAcl(params, function(err, data) {
	  if (err) console.log(err, err.stack); // an error occurred
	  else	{
		  console.log('get bucket acl succeed');
		  console.log(data);
		  deferred.resolve(data);
	  }
	});
	return deferred.promise;	
}

function getObjectAcl (params, s3)
{
	var deferred = Q.defer();
	s3.getObjectAcl(params, function(err, data) {
		  if (err) {
			  console.log(err, err.stack);
			  deferred.resolve();
		  }// an error occurred
		  else     {
			  console.log('get object acl succeed');
			  console.log(data);
			  deferred.resolve(data);// successful response
		  }
		});
	return deferred.promise;	
}


function putObjectAcl(params, s3)
{
	var deferred = Q.defer();
	s3.putObjectAcl(params, function(err, data) {
		  if (err) {
			  console.log(err, err.stack);
			  deferred.resolve();
		  }
		  else     {
			  console.log('put object acl succeed');
			  console.log(data);
			  deferred.resolve(data);		  
		  }
		});
	return deferred.promise;	

}